src/server/gdb_server.c

```diff
@@ -1785,7 +1785,13 @@ static int gdb_breakpoint_watchpoint_packet(struct connection *connection,
 		case 0:
 		case 1:
 			if (packet[0] == 'Z') {
-				retval = breakpoint_add(target, address, size, bp_type);
+				struct target *bp_target = target;
+				if (target->rtos && bp_type == BKPT_SOFT) {
+					bp_target = rtos_swbp_target(target, address, size, bp_type);
+					if (!bp_target)
+						return ERROR_FAIL;
```

Comment:
It would be good to re-test the code with a conventional RTOS (something else than `hwthread`, for example FreeRTOS).

There is at least one FreeRTOS-based test in the `riscv-tests` repo, @timsifive will know the details: 
- https://github.com/riscv-software-src/riscv-tests/blob/master/debug/gdbserver.py#L1810

Reply:
If an RTOS doesn't define swbp_target(), then rtos_swbp_target() will return the original target, so you only get NULL if the RTOS implemented swbp_target() and decided to return NULL.